### PR TITLE
Handle DASC storage round trips and imports

### PR DESCRIPTION
**Pull request by kaix-nv**
## Summary

Follow-up to #2380 that addresses its complete automated review pass:

- bound policy compatibility against FP16 and BF16 storage rounding even when checkpoint values are reloaded into a wider FP32 tensor
- expose optional-framework resolution failures as warnings and report fully qualified supported class paths
- distinguish unsupported ordinary GDN subclasses from models with no GDN layer
- resolve GDN modules once per analysis or validation operation
- guard installed Megatron and Transformers class paths against dependency path drift

This PR is intentionally stacked on #2380 because repository rules protect a PR head branch after creation.

## Validation

- focused DASC suite: 20 passed, 1 optional Megatron skip; installed Transformers path passed
- DASC plus weight sparsity plus attention sparsity compatibility suite: 131 passed, 1 optional skip
- DASC package coverage: 400/400 statements, 100%
- full pre-commit on touched files: passed
- BF16 storage to FP32 reload regression: passed

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved warnings and error messages for unavailable, invalid, or unsupported dynamic modules.
  - Improved sparsity storage rounding for FP16, BF16, and other tensor data types.
  - Added validation to reject non-floating-point decay parameters.
  - Improved checkpoint reload handling for BF16-rounded models.

- **Tests**
  - Expanded coverage for module discovery, import failures, invalid paths, unsupported subclasses, and dtype-specific checkpoint behavior.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -166,12 +210,19 @@ def _decay_parameters(modules: dict[str, nn.Module]) -> list[dict[str, object]]:
...
+    values = tensor.detach().to(device="cpu", dtype=torch.float64).abs()
+    radius = torch.zeros_like(values)
+    for dtype in (tensor.dtype, *_SUPPORTED_STORAGE_DTYPES):
+        dtype_info = torch.finfo(dtype)
+        unit_roundoff = dtype_info.eps / 2.0
+        candidate = (
+            values * (unit_roundoff / (1.0 - unit_roundoff)) + dtype_info.tiny * dtype_info.eps
+        )
+        radius = torch.maximum(radius, candidate)
+    return radius
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:225]: Addressed in #2383. DASCConfig and DASCPolicy now record decay_parameter_storage_dtype, defaulting to strict float32 for backward-compatible parsing. Bounds compose only that declared storage cast and any distinct live-dtype materialization cast. Provenance hashing uses the same declared dtype. Tests cover strict 1e-4 horizon tampering, BF16 to FP32 reload, FP16 to BF16, BF16 to FP16, and legacy metadata.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 2]
```diff
@@ -166,12 +210,19 @@ def _decay_parameters(modules: dict[str, nn.Module]) -> list[dict[str, object]]:
...
 def _storage_rounding_radius(tensor: torch.Tensor) -> torch.Tensor:
-    """Bound one cast-to-storage rounding step around the represented tensor values."""
-    values = tensor.detach().to(device="cpu", dtype=torch.float64)
-    dtype_info = torch.finfo(tensor.dtype)
-    unit_roundoff = dtype_info.eps / 2.0
-    subnormal_slack = dtype_info.tiny * dtype_info.eps
-    return values.abs() * (unit_roundoff / (1.0 - unit_roundoff)) + subnormal_slack
+    """Bound one FP16/BF16 storage cast even after values are reloaded in a wider dtype."""
+    if not tensor.dtype.is_floating_point:
+        raise ApplyModeError("DASC decay parameters must use a floating-point dtype")
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:215]: Addressed in #2383. Floating-point dtype validation now occurs in compute_gdn_decay_horizons before conversion or mask derivation, so integer decay tensors produce the direct dtype error independent of the resulting mask.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 3]
```diff
@@ -99,22 +116,37 @@ def _is_gdn_module(module: nn.Module) -> bool:
...
+        ]
+        if unsupported_subclasses:
+            raise ApplyModeError(
+                "DASC found GDN subclasses that are not ModelOpt dynamic modules at: "
+                f"{', '.join(unsupported_subclasses)}; use an exact supported class"
+            )
+        supported = ", ".join(
+            f"{module_name}.{class_name}" for module_name, class_name in _SUPPORTED_GDN_CLASS_PATHS
+        )
         raise ApplyModeError(f"DASC found no supported GDN modules; expected one of: {supported}")
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:140]: Addressed in #2383. The subclass message now says to convert with ModelOpt or use a supported class directly. A separate diagnostic detects an exact or ModelOpt-dynamic supported identity missing A_log or dt_bias tensors, with regression coverage.

File: tests/unit/torch/sparsity/state_sparsity/test_dasc.py [hunk 4]
```diff
@@ -263,21 +263,42 @@ def test_supported_class_resolution_uses_imported_module_identities(monkeypatch)
...
+
+
+@pytest.mark.parametrize(("module_name", "class_name"), dasc_policy._SUPPORTED_GDN_CLASS_PATHS)
+def test_declared_gdn_paths_resolve_when_framework_is_installed(module_name, class_name):
+    """Guard supported identities against upstream dependency path drift."""
+    root_module = module_name.partition(".")[0]
+    if dasc_policy.importlib.util.find_spec(root_module) is None:
+        pytest.skip(f"optional framework {root_module!r} is not installed")
+    module = dasc_policy.importlib.import_module(module_name)
+    assert issubclass(getattr(module, class_name), nn.Module)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply tests/unit/torch/sparsity/state_sparsity/test_dasc.py:301]: Addressed in #2383. The real-path guard now uses pytest.importorskip on the framework root. An unimportable optional root skips, while an importable root with a moved leaf path or class still fails.
